clang/lib/CIR/CodeGen/CIRGenClass.cpp

```diff
@@ -1109,7 +1109,17 @@ void CIRGenFunction::buildDestructorBody(FunctionArgList &Args) {
   // in fact emit references to them from other compilations, so emit them
   // as functions containing a trap instruction.
   if (DtorType != Dtor_Base && Dtor->getParent()->isAbstract()) {
-    llvm_unreachable("NYI");
+    builder.create<mlir::cir::TrapOp>(getLoc(Dtor->getLocation()));
+    // cir.trap is a terminator. The corresponding IRGen code does this:
+    // builder.clearInsertionPoint();
+    // However, CIRGenFunction::generateCode performs the following check, which
+    // fails if we do the same thing here:
+    // assert(builder.getInsertionBlock() && "Should be valid");
+    // IRGen has no equivalent assert, but for now we create a block to have a
+    // valid insertion point.
+    // TODO(cir): Should the CIRGenFunction::generateCode assert be relaxed?
```

Comment:
I'd appreciate advice on this.

Reply:
MLIR builder asserts the insertion point to be valid when inserting operations while LLVMIR builder does not. Thus CodeGen could clear the insertion point after emitting an terminator but CIRGen could not.

The assert statement in `CIRGenFunction::generateCode` ensures that the cleanup of the function-level lexical scope could be successfully generated in `LexicalScope::~LexicalScope`. Note that the cleanup also includes the return operation of the function. Since we currently have no means to track whether all paths to the end of a function is non-returning, the return op has to be emitted, thus the insertion point has to be kept alive by now.

Reply:
The fact that OG codegen (IRGen) clears insertion points is sometimes a bit brittle (used to signal some behavior that could be tracked more clearly). When emitting initializer for globals in CIR we sometimes also have no insertion point because those don't come from a regular function, and that's fine over there.

Just remove the overall question in form of comment and add a note saying that OG checks the insertion point but we do not.